File: packages/davinci/ios/Factory/DaVinciClientFactory.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingDavinci
import PingLogger
import PingOidc
import PingStorage
import RNPingCore
#if canImport(PingOneProtect)
import PingOneProtect
#endif

/// Builds native DaVinci workflow instances from parsed JS payloads.
final class DaVinciClientFactory {
  /// Resolved OIDC configuration used to construct the native OIDC module.
  private struct ResolvedOidcConfig: Sendable {
    let discoveryEndpoint: String
    let clientId: String
    let redirectUri: String
    let par: Bool?
    let scopes: [String]
    let storageId: String?
    let signOutRedirectUri: String?
    let loginHint: String?
    let nonce: String?
    let state: String?
    let prompt: String?
    let display: String?
    let uiLocales: String?
    let acrValues: String?
    let refreshThreshold: Int64?
    let additionalParameters: [String: String]
  }

  /// Builds a DaVinci workflow from parsed configuration.
  ///
  /// - Parameter payload: Parsed DaVinci payload.
  /// - Returns: Configured native DaVinci instance.
  /// - Throws: `DaVinciBridgeError.argument` when shared registries cannot resolve handles.
  func build(_ payload: DaVinciClientPayload) async throws -> DaVinci {
    let resolvedLogger = await resolveLoggerFromCore(payload.loggerId)
    let resolvedOidcConfig = resolveOidcConfig(payload)
    let oidcStorage = try await Self.buildOidcStorageDelegate(resolvedOidcConfig.storageId)

    return DaVinci.createDaVinci { config in
      if let resolvedLogger {
        config.logger = resolvedLogger
      }
      if let timeoutSeconds = Self.timeoutSeconds(from: payload.timeout) {
        config.timeout = timeoutSeconds
      }

      config.module(PingDavinci.OidcModule.config) { oidcConfig in
        oidcConfig.discoveryEndpoint = resolvedOidcConfig.discoveryEndpoint
        oidcConfig.clientId = resolvedOidcConfig.clientId
        oidcConfig.redirectUri = resolvedOidcConfig.redirectUri
        if let par = resolvedOidcConfig.par {
          oidcConfig.par = par
        }
        oidcConfig.scopes = Set(resolvedOidcConfig.scopes)

        if let loginHint = resolvedOidcConfig.loginHint {
          oidcConfig.loginHint = loginHint
        }
        if let nonce = resolvedOidcConfig.nonce {
          oidcConfig.nonce = nonce
        }
        if let state = resolvedOidcConfig.state {
          oidcConfig.state = state
        }
        if let prompt = resolvedOidcConfig.prompt {
          oidcConfig.prompt = prompt
        }
        if let display = resolvedOidcConfig.display {
          oidcConfig.display = display
        }
        if let uiLocales = resolvedOidcConfig.uiLocales {
          oidcConfig.uiLocales = uiLocales
        }
        if let acrValues = resolvedOidcConfig.acrValues {
          oidcConfig.acrValues = acrValues
        }
        if let refreshThreshold = resolvedOidcConfig.refreshThreshold {
          oidcConfig.refreshThreshold = refreshThreshold
        }
        if !resolvedOidcConfig.additionalParameters.isEmpty {
          oidcConfig.additionalParameters = resolvedOidcConfig.additionalParameters
        }
        if let oidcStorage {
          oidcConfig.storage = oidcStorage
        }
        if resolvedOidcConfig.signOutRedirectUri != nil {
          // TODO-SDK-PARITY: OidcClientConfig currently does not expose
          // `signOutRedirectUri` on iOS. Android supports this field.
        }
      }

      #if canImport(PingOneProtect)
      if let protect = payload.protect {
        config.module(ProtectLifecycleModule.config) { protectConfig in
          protectConfig.envId = protect.envId
          protectConfig.isBehavioralDataCollection = protect.isBehavioralDataCollection
          protectConfig.isLazyMetadata = protect.isLazyMetadata
          if let customHost = protect.customHost {
            protectConfig.customHost = customHost
          }
          protectConfig.isConsoleLogEnabled = protect.isConsoleLogEnabled
          if !protect.deviceAttributesToIgnore.isEmpty {
            protectConfig.deviceAttributesToIgnore = protect.deviceAttributesToIgnore
          }
          protectConfig.pauseBehavioralDataOnSuccess = protect.pauseBehavioralDataOnSuccess
          protectConfig.resumeBehavioralDataOnStart = protect.resumeBehavioralDataOnStart
        }
      }
      #endif
    }
  }

  /// Resolves the direct OIDC configuration for native module construction.
  ///
  /// - Parameter payload: Parsed DaVinci payload.
  /// - Returns: Resolved DaVinci OIDC configuration.
  private func resolveOidcConfig(_ payload: DaVinciClientPayload) -> ResolvedOidcConfig {
    let oidc = payload.oidc
    return ResolvedOidcConfig(
      discoveryEndpoint: oidc.discoveryEndpoint,
      clientId: oidc.clientId,
      redirectUri: oidc.redirectUri,
      par: oidc.par,
      scopes: oidc.scopes,
      storageId: oidc.storageId,
      signOutRedirectUri: oidc.signOutRedirectUri,
      loginHint: oidc.loginHint,
      nonce: oidc.nonce,
      state: oidc.state,
      prompt: oidc.prompt,
      display: oidc.display,
      uiLocales: oidc.uiLocales,
      acrValues: oidc.acrValues,
      refreshThreshold: oidc.refreshThreshold,
      additionalParameters: oidc.additionalParameters
    )
  }

  /// Resolve a native logger from the shared Core logger registry.
  ///
  /// - Parameter loggerId: Logger handle identifier from JS.
  /// - Returns: Native logger instance, or `nil` when missing/invalid.
  private func resolveLoggerFromCore(_ loggerId: String?) async -> Logger? {
    guard let loggerId, !loggerId.isEmpty else {
      return nil
    }
    guard let handle = await CoreRuntime.loggerRegistry.resolve(loggerId) as? LoggerHandleContract else {
      return nil
    }
    return handle.nativeLogger as? Logger
  }

  /// Builds an OIDC storage delegate from a registered storage id.
  ///
  /// - Parameter storageId: OIDC storage identifier from JS.
  /// - Returns: Storage delegate, or `nil` when no storage id is provided.
  /// - Throws: `DaVinciBridgeError.argument` when a provided id cannot be resolved.
  private static func buildOidcStorageDelegate(_ storageId: String?) async throws -> StorageDelegate<Token>? {
    guard let storageId, !storageId.isEmpty else {
      return nil
    }
    guard let config = await CoreRuntime.oidcStorageConfigRegistry.resolve(storageId) as? StorageConfigHandleContract else {
      throw DaVinciBridgeError.argument("No OIDC storage config registered for id=\(storageId)")
    }
    let account = config.account ?? "com.pingidentity.rndavinci.storage"
    let encryptorEnabled = config.encryptor ?? true
    let encryptor: Encryptor = {
      if encryptorEnabled, let secured = SecuredKeyEncryptor() {
        return secured
      }
      return NoEncryptor()
    }()
    return KeychainStorage<Token>(
      account: account,
      encryptor: encryptor,
      cacheStrategy: parseCacheStrategy(cacheable: config.cacheable, rawStrategy: nil)
    )
  }

  /// Maps storage config payload values to PingStorage cache strategy.
  ///
  /// Supports both the current `cacheable` boolean and future `cacheStrategy` string.
  ///
  /// - Parameters:
  ///   - cacheable: Storage cacheable flag.
  ///   - rawStrategy: Optional explicit cache strategy string.
  /// - Returns: Native cache strategy value.
  private static func parseCacheStrategy(cacheable: Bool?, rawStrategy: String?) -> CacheStrategy {
    if let rawStrategy = rawStrategy?.lowercased() {
      switch rawStrategy {
      case "cache":
        return .CACHE
      case "cache_on_failure":
        return .CACHE_ON_FAILURE
      case "no_cache":
        return .NO_CACHE
      default:
        break
      }
    }

    if let cacheable {
      return cacheable ? .CACHE_ON_FAILURE : .NO_CACHE
    }

    return .NO_CACHE
  }

  /// Converts JS timeout milliseconds into DaVinci timeout seconds.
  ///
  /// - Parameter timeoutMs: Timeout in milliseconds.
  /// - Returns: Timeout in seconds for DaVinci configuration.
  private static func timeoutSeconds(from timeoutMs: Int64?) -> TimeInterval? {
    guard let timeoutMs else {
      return nil
    }
    if timeoutMs <= 0 {
      return nil
    }
    return max(1, ceil(Double(timeoutMs) / 1000.0))
  }
}

File: packages/davinci/android/src/main/java/com/pingidentity/rndavinci/factory/DaVinciClientFactory.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rndavinci.factory

import com.pingidentity.davinci.DaVinci
import com.pingidentity.davinci.module.Oidc
import com.pingidentity.oidc.OidcClientConfig
import com.pingidentity.orchestrate.Workflow
import com.pingidentity.rncore.logger.LoggerHandleContract
import com.pingidentity.rncore.registry.Registry
import com.pingidentity.rncore.storage.StorageConfigHandleContract
import com.pingidentity.rndavinci.config.DaVinciClientPayload
import com.pingidentity.storage.CacheStrategy
import com.pingidentity.storage.EncryptedDataStoreStorageConfig
// ProtectLifecycle is compileOnly (provided by rn-protect); guarded at runtime with NoClassDefFoundError.
import com.pingidentity.protect.ProtectLifecycle

/**
 * Builds native DaVinci workflow instances from parsed JS payloads.
 *
 * @param oidcStorageRegistry Registry used to resolve OIDC storage handles from JS.
 * @param loggerRegistry Registry used to resolve logger handles from JS.
 */
internal class DaVinciClientFactory(
    private val oidcStorageRegistry: Registry,
    private val loggerRegistry: Registry,
) {

    private data class ResolvedOidcConfig(
        val discoveryEndpoint: String,
        val clientId: String,
        val redirectUri: String,
        val scopes: List<String>,
        val par: Boolean?,
        val storageId: String?,
        val signOutRedirectUri: String?,
        val loginHint: String?,
        val nonce: String?,
        val state: String?,
        val prompt: String?,
        val display: String?,
        val uiLocales: String?,
        val acrValues: String?,
        val refreshThreshold: Long?,
        val additionalParameters: Map<String, String>,
    )

    private fun resolveOidcConfig(payload: DaVinciClientPayload): ResolvedOidcConfig {
        val oidc = payload.oidc
        return ResolvedOidcConfig(
            discoveryEndpoint = oidc.discoveryEndpoint,
            clientId = oidc.clientId,
            redirectUri = oidc.redirectUri,
            scopes = oidc.scopes,
            par = oidc.par,
            storageId = oidc.storageId,
            signOutRedirectUri = oidc.signOutRedirectUri,
            loginHint = oidc.loginHint,
            nonce = oidc.nonce,
            state = oidc.state,
            prompt = oidc.prompt,
            display = oidc.display,
            uiLocales = oidc.uiLocales,
            acrValues = oidc.acrValues,
            refreshThreshold = oidc.refreshThreshold,
            additionalParameters = oidc.additionalParameters,
        )
    }

    /**
     * Build a DaVinci workflow from parsed configuration.
     *
     * @param payload Parsed DaVinci payload.
     * @return Configured native workflow instance.
     * @throws IllegalArgumentException if payload values are invalid for native SDK setup.
     */
    fun build(payload: DaVinciClientPayload): Workflow {
        val resolvedOidcConfig = resolveOidcConfig(payload)
        val resolvedLogger = resolveLogger(payload.loggerId)
        return DaVinci {
            resolvedLogger?.let { logger = it }
            payload.timeout?.let { timeout = it }

            module(Oidc) {
                discoveryEndpoint = resolvedOidcConfig.discoveryEndpoint
                clientId = resolvedOidcConfig.clientId
                redirectUri = resolvedOidcConfig.redirectUri
                scopes = resolvedOidcConfig.scopes.toMutableSet()
                resolvedOidcConfig.par?.let { par = it }
                resolvedOidcConfig.signOutRedirectUri?.let { signOutRedirectUri = it }
                resolvedOidcConfig.loginHint?.let { loginHint = it }
                resolvedOidcConfig.nonce?.let { nonce = it }
                resolvedOidcConfig.state?.let { state = it }
                resolvedOidcConfig.prompt?.let { prompt = it }
                resolvedOidcConfig.display?.let { display = it }
                resolvedOidcConfig.uiLocales?.let { uiLocales = it }
                resolvedOidcConfig.acrValues?.let { acrValues = it }
                resolvedOidcConfig.refreshThreshold?.let { refreshThreshold = it }
                if (resolvedOidcConfig.additionalParameters.isNotEmpty()) {
                    additionalParameters = resolvedOidcConfig.additionalParameters
                }
                applyOidcStorageIfPresent(resolvedOidcConfig.storageId)
            }

            payload.protect?.let { protect ->
                try {
                    module(ProtectLifecycle) {
                        protect.envId?.let { envId = it }
                        isBehavioralDataCollection = protect.isBehavioralDataCollection
                        isLazyMetadata = protect.isLazyMetadata
                        protect.customHost?.let { customHost = it }
                        isConsoleLogEnabled = protect.isConsoleLogEnabled
                        if (protect.deviceAttributesToIgnore.isNotEmpty()) {
                            deviceAttributesToIgnore = protect.deviceAttributesToIgnore
                        }
                        pauseBehavioralDataOnSuccess = protect.pauseBehavioralDataOnSuccess
                        resumeBehavioralDataOnStart = protect.resumeBehavioralDataOnStart
                    }
                } catch (e: NoClassDefFoundError) {
                    logger?.e("modules.protect was configured but the PingOne Protect SDK is not on the classpath; ProtectLifecycle was NOT registered. Add com.pingidentity.sdks:protect.", e)
                }
            }
        }
    }

    /**
     * Resolves a native logger from the shared Core logger registry.
     *
     * @param id Logger handle identifier from JS.
     * @return Native logger instance, or null when missing/invalid.
     */
    private fun resolveLogger(id: String?): com.pingidentity.logger.Logger? {
        id ?: return null
        if (id.isBlank()) return null
        val handle = loggerRegistry.resolve(id) as? LoggerHandleContract ?: return null
        return handle.nativeLogger as? com.pingidentity.logger.Logger
    }

    /**
     * Applies OIDC storage configuration when a storage id is present.
     *
     * @param storageId Optional OIDC storage handle id.
     * @throws IllegalArgumentException if the storage id cannot be resolved.
     */
    private fun OidcClientConfig.applyOidcStorageIfPresent(storageId: String?) {
        if (storageId.isNullOrBlank()) return
        val storageConfig = oidcStorageRegistry.resolve(storageId) as? StorageConfigHandleContract
            ?: throw IllegalArgumentException("No OIDC storage config registered for id=$storageId")
        storage {
            applyStorageConfig(storageConfig)
        }
    }

    /**
     * Applies storage configuration fields to native encrypted datastore storage settings.
     *
     * @param config Parsed storage configuration from the shared storage registry.
     */
    private fun EncryptedDataStoreStorageConfig.applyStorageConfig(config: StorageConfigHandleContract) {
        config.fileName?.let { fileName = it }
        config.keyAlias?.let { keyAlias = it }
        config.strongBoxPreferred?.let { strongBoxPreferred = it }
        config.cacheStrategy?.let { cacheStrategy = parseCacheStrategy(it) }
    }

    /**
     * Maps wire cache strategy strings to native enum values.
     *
     * @param rawValue Raw cache strategy value from JavaScript.
     * @return Native cache strategy enum.
     */
    private fun parseCacheStrategy(rawValue: String): CacheStrategy {
        return when (rawValue.lowercase()) {
            "cache_on_failure" -> CacheStrategy.CACHE_ON_FAILURE
            "no_cache" -> CacheStrategy.NO_CACHE
            "cache" -> CacheStrategy.CACHE
            else -> runCatching { CacheStrategy.valueOf(rawValue) }.getOrDefault(CacheStrategy.NO_CACHE)
        }
    }
}
